patches/server/0969-New-specific-event-watching-command.patch

```diff
@@ -0,0 +1,196 @@
++        JsonWriter jsonWriter = new JsonWriter(stringWriter);
++        jsonWriter.setIndent(" ");
++        jsonWriter.setLenient(false);
++        Streams.write(array, jsonWriter);
++
++        String[] unformatted = event.split("\\.");
++        String eventName = unformatted[unformatted.length - 1];
++
++        Path parent = PARENT.resolve(eventName);
++        Path path = parent.resolve("event-" + eventName + "-" +
++            FORMATTER.format(LocalDateTime.now()) + ".txt");
++        Files.createDirectories(parent);
++        Path file = Files.createFile(path);
++
++        try (PrintStream out = new PrintStream(Files.newOutputStream(file), false, StandardCharsets.UTF_8)) {
++            out.print(stringWriter);
++        }
++        sender.sendMessage(text("Successfully dumped event information to " + file, GREEN));
++        watching.remove(event);
++    }
++
++    /**
++     * EventWrapper subclass to apply listener information to our Json object
++     */
++    private static final class Wrapper extends EventWrapper {
++        private final Listener listener;
++
++        public Wrapper(Event event, @Nullable Listener listener) {
++            super(event);
++            this.listener = listener;
++        }
++
++        public JsonObject serialize() {
++            JsonObject obj = super.serialize();
++            if (obj == null)
++                return null;
++            obj.addProperty("listener", listener == null ? "Initial Call" : listener.getClass().getCanonicalName());
++            return obj;
++        }
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+index dfb91cb8e19e0159e0eba9367ca7396125da6d1c..48dd9149110a3e7548a610eb80d6d301ba6121e2 100644
+--- a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
++++ b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+@@ -5,6 +5,7 @@ import com.destroystokyo.paper.event.server.ServerExceptionEvent;
+ import com.destroystokyo.paper.exception.ServerEventException;
+ import com.google.common.collect.Sets;
+ import io.papermc.paper.command.subcommands.EventVerboseCommand;
++import io.papermc.paper.command.subcommands.EventWatcherCommand;
+ import org.bukkit.Server;
+ import org.bukkit.Warning;
+ import org.bukkit.event.Event;
+@@ -43,6 +44,7 @@ class PaperEventManager {
+             throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
+         }
+         EventVerboseCommand.tryLogEvent(event);
++        EventWatcherCommand.tryLog(event, null);
+ 
+         HandlerList handlers = event.getHandlers();
+         RegisteredListener[] listeners = handlers.getRegisteredListeners();
+@@ -54,6 +56,7 @@ class PaperEventManager {
+ 
+             try {
+                 registration.callEvent(event);
++                EventWatcherCommand.tryLog(event, registration.getListener());
```

Comment:
Should also log the event executor, as that often has info about the actual method being called